### DISCUSSION TITLE
NCM On windows 10

### DISCUSSION THREAD

**Discussion by fabien251** [hathach/tinyusb#2727] [Q&A]
Hi everyone,

I just wrote a new program using TinyUSB with an STM32L476RGT6. I'm trying to implement a web page with LwIP that is accessible via USB using NCM from TinyUSB with the net_lwip_webserver. It works very well on Windows 11, but my device is not recognized on Windows 10 (error 28) ! Does anyone know why?

Fabien

**Comment by HiFiPhile** [comment]
On windows 10 you need to install the driver manually, choose Microsoft -> UsbNcm Host Device

**Reply by fabien251** [reply]
Thanks for your reply, @HiFiPhile.

However, the problem is that I have an NCM device that works on both my Windows 10 and Windows 11 laptop, so I already have the driver on my Windows 10 machine. I imagine there must be a subtlety somewhere in the USB descriptor that I am not aware of.

**Reply by HiFiPhile** [reply]
I believe the issue is in Microsoft's side, Windows 10's NCM driver is not configured to do device binding automatically.

**Reply by fabien251** [reply]
Thanks for your reply, @HiFiPhile.
That might be the case, but I still don't understand why some devices are detected on Windows 10 and others are not. Do you know what could be causing this issue? I know that the working one uses an RP2040 and TinyUSB—could the microcontroller be related to this problem?

**Reply by HiFiPhile** [reply]
I don't think MCU make a difference. Are you sure it was recognized on first plug in ?  Maybe make a descriptor dump to see the difference.

**Reply by fabien251** [reply]
Yes, it was recognized on the first plug-in. As I observed on Wireshark, Windows 10 doesn't send the GET_STATUS_REQUEST, unlike my Windows 11. I'm a beginner with USB, so any guidance would be appreciated.

**Reply by offlinemark** [reply]
Hey @fabien251 - On Win10, NCM devices are required to send a Microsoft OS descriptor for the NCM driver to autoload. Not sure exactly what happens at the USB level, but on Linux, this is done via the below in configfs.

```
echo 1       > os_desc/use
echo 0xbc    > os_desc/b_vendor_code
echo MSFT100 > os_desc/qw_sign
```

```
echo WINNCM   > functions/ncm.usb0/os_desc/interface.ncm/compatible_id
```

See https://github.com/vpelletier/python-functionfs/issues/20.

Windows also caches info about devices and may not request OS descriptors if a device previously didn't have any. You can clear this registry key to reset this cache.

```
HLKM\SYSTEM\CurrentControlSet\Control\UsbFlags\vvvvpppprrrrr
```

https://learn.microsoft.com/en-us/windows-hardware/drivers/usbcon/microsoft-defined-usb-descriptors

HTH!

**Reply by HiFiPhile** [reply]
Hi, MS OS2.0 descriptor has been added in #2829 in order to auto load driver on Windows 10.
